File: word-break-ii.cc
// Word Break II
class Solution {
  string s;
  vector<string> dict, ans;
  vector<int> parts;
  void dfs(int n) {
    if (n == 0) {
      string ss;
      for (int j = 0, i = parts.size(); i--; ) {
        if (j) ss += ' ';
        ss.insert(ss.size(), &s[j], parts[i]);
        j += parts[i];
      }
      ans.push_back(ss);
    } else
      for (auto &w: dict)
        if (w.size() <= n && s.compare(n-w.size(), w.size(), w) == 0) {
          parts.push_back(w.size());
          dfs(n-w.size());
          parts.pop_back();
        }
  }
public:
  vector<string> wordBreak(string s, vector<string>& wordDict) {
    this->s = move(s);
    dict = move(wordDict);
    ans.clear();
    parts.clear();
    dfs(this->s.size());
    return ans;
  }
};

File: word-break-ii.rs
impl Solution {
    pub fn word_break(s: String, word_dict: Vec<String>) -> Vec<String> {
        struct Sol<'a> { dict: &'a [String], ans: Vec<String>, parts: Vec<&'a str> }
        impl<'a> Sol<'a> {
            fn dfs(self: &mut Self, s: &str) {
                if s.is_empty() {
                    self.ans.push(self.parts.join(" "));
                    return;
                }
                for w in self.dict.iter() {
                    if s.starts_with(w) {
                        self.parts.push(w);
                        self.dfs(&s[w.len()..]);
                        self.parts.pop();
                    }
                }
            }
        }
        let mut sol = Sol { dict: &word_dict, ans: vec![], parts: vec![] };
        sol.dfs(&s);
        sol.ans
    }
}
